### ISSUE TITLE
buildRustCrate fails with anyhow crate

### ISSUE THREAD

**Issue by kolloch** [NixOS/nixpkgs#74071]
**Describe the bug**
`buildRustCrate` fails to compile the anyhow crate because 

1. Anyhow creates a `lib.rs` file for a probe in its build.rs.
2. This `lib.rs` is copied over the real anyhow lib.rs so that the probe is compiled instead of the real lib.

This second step seems to be performed differently by cargo if at all. I can imagine that this step was inspired by [this section](https://doc.rust-lang.org/cargo/reference/build-scripts.html#case-study-code-generation).

The corresponding nix source seems to be [here](https://github.com/NixOS/nixpkgs/blob/bad07dfac57c36dff1c7b3fd4020fea806d7be93/pkgs/build-support/rust/build-rust-crate/configure-crate.nix#L142-L149)

**To Reproduce**
1. 
```shell
    $ cat >anyhow.nix
{ pkgs? import <nixpkgs> { config = {}; },
  lib? pkgs.lib,
  callPackage? pkgs.callPackage,
  stdenv? pkgs.stdenv,
  buildRustCrate? pkgs.buildRustCrate}:

buildRustCrate {
    crateName = "anyhow";
    version = "1.0.23";
    edition = "2018";
    sha256 = "109xr34615f5z3dv5wvwlfmygzr8k72y67f6b6dxs71bzrjgrjg8";
}
```

2. `nix-build anyhow.nix`

**Expected behavior**
The original src/lib.rs of anyhow should not be overwritten by the probe.

**Additional context**
This is where the `lib.rs` is written from the `build.rs` script:

https://github.com/dtolnay/anyhow/blob/1.0.23/build.rs#L47-L48

This is the original issue where this came up:

https://github.com/kolloch/crate2nix/issues/39

**Metadata**
Please run `nix run nixpkgs.nix-info -c nix-info -m` and paste the result.

```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.84, NixOS, 19.09.1320.4ad6f1404a8 (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(root): `"nixos-19.09.1320.4ad6f1404a8, nixos-18.03-18.03.133402.cb0e20d6db9, nixos-unstable-20.03pre202088.e89b21504f3"`
 - channels(peter): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by andir** [comment]
This really seems smells like we should remove the overlaying code. `OUT_DIR` should be all that is required after the build.rs script has been run. I'll give this a spin in a bit.

**Comment by andir** [comment]
This should have been fixed by #78706
